### ISSUE TITLE
vehicle access

### ISSUE THREAD

**Issue by hungerburg** [valhalla/valhalla#3882]
Valhalla routes bicycle, where vehicle=forestry is tagged on a track|path. I have searched a bit and found taginfo.json. Vehicle is not mentioned there. Maybe it is missing somewhere?

PS: I should say, I used valhalla from the OSM site. And, maybe the access tags in place are used, Idk. If the directions said: push the bike there, I would know.

**Comment by kevinkreiser** [comment]
Yep our lua processing doesn't care about the vehicle tag at all. For bike we do use access tags and bike specific tags. Can you post a link to an example way and tell us what you expect to happen just so we are sure we understand properly? Thabk you!

**Comment by nilsnolde** [comment]
Wasn't even really aware that we don't process that. I think it'd generally make sense to add `vehicle` to our graph.lua, it exists on [> 300k ways](https://taginfo.openstreetmap.org/keys/vehicle#values). In Europe that's a frequently used tag, lots of countries have a dedicated sign to forbid vehicle access, often allowing bicycles. While that's the same as `motor_vehicle=no`, often even bicycles aren't allowed, so it's shorter with a `vehicle=no`.

**Comment by hungerburg** [comment]
Let's visit a friend at Poltenhütte, starting from Sillufer. There is a maze of paths, so use a router https://www.openstreetmap.org/directions?engine=fossgis_valhalla_bicycle&route=47.25342%2C11.40457%3B47.24804%2C11.41301#map=16/47.2485/11.4115

Attached screenshot, three pushing sections marked. Vehicle=forestry is country-wide default access in Austria on paths/tracks in the woods. Some influential mappers say, the router should infer from environment. I doubt that will ever be possible reliably. It gets tagged though, on paths and tracks. Sometimes an extra sign is mounted below the no driving sign "Also applies to bicycles" (they are vehicles.)

![Screenshot-Openstreetmap-Valhalla-Bicycle](https://user-images.githubusercontent.com/586972/210059606-7132ccea-01ca-41ae-912a-aa187a87ba59.png)

**Comment by nilsnolde** [comment]
Thanks! 

I didn’t have a look but is it mentioned somewhere in the OSM wiki for which specific highway tags that rule generally applies? We generally have the possibility of respecting such implicit, country-specific default behavior with Valhalla, but again, also didn’t look at the code to see what’s already applied to Austria, if any.

**Comment by hungerburg** [comment]
@nilsnolde I hope I understood correctly: I'd say, controlling for vehicle=forestry tags makes sense on highway=track|path (In order of diminishing returns). There it is the most correct tagging to map the law of the land. On highway=service|unclassified|residential it need not be a tagging error, sometimes there are signs on the ground that make wonder. Returns are minimal though - https://overpass-turbo.eu/s/1py4

On the implicit rule - here it gets complicated - For a start, the law does not restrict access, instead it allows access, namely on foot (including ski, but not bicycle etc.) to what is to a very large extent private property, where there is wood according to definition in §1a [Forstgesetz](https://www.ris.bka.gv.at/GeltendeFassung.wxe?Abfrage=Bundesnormen&Gesetzesnummer=10010371) - [Austrian Forest Act](https://info.bml.gv.at/dam/jcr:de7d5f34-a758-45b0-a22a-30c4598c910d/Forstgesetz%20idF.%20BGBl.%20I%20Nr.%2056-2016%20englisch.doc) - This might differ too much from what gets mapped landuse=forest or natural=wood in openstreetmap?

**Comment by nilsnolde** [comment]
Also interesting what pedestrian routing does in comparison:
https://www.openstreetmap.org/directions?engine=fossgis_valhalla_foot&route=47.25342%2C11.40457%3B47.24804%2C11.41301#map=17/47.24966/11.40883

We seem to have quite a few issues with the tagging of those ways.

![image](https://user-images.githubusercontent.com/25637358/210323297-50db7c8e-c0e8-4b1c-9e1d-8460825b9bab.png)

**Comment by hungerburg** [comment]
Hello Nils, The cycling directions are just as good for walking there. The hiking route Via Romea runs here https://hiking.waymarkedtrails.org/#?map=17.0/47.25/11.41

Two ideas: Make secondary highway without sidewalk mapped and paths where trail_visibility is bad much more expensive.

Indeed, Lanser Steig is steep. There is no better alternative - Why does pedestrian shy from tracks so much? [(see Oberer Gang Weg)](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.4168119430542%2C47.24959995235856%2C11.413024663925173%2C47.2480596184114)

The no-pushing, at least not for legal, cycle itinerary would go [all over Vill](https://valhalla.openstreetmap.de/directions?profile=bicycle&wps=11.40435576438904%2C47.253000909324136%2C11.405332088470459%2C47.2426334990063%2C11.412987112998964%2C47.248052335402186). Mind you, all tracks in the woods in Austria are bicycle=dismount, not just the ones where mapping says so.

Maybe valhalla is just not meant for such "workout" like routes? The duration estimates are fine tough!

**Comment by nilsnolde** [comment]
`trail_visibility` is also a new one to me :smile: would be nice for bike @dnesbitt61 :)

> Why does pedestrian shy from tracks so much?

You can favor elevation or avoid it with `use_hills`. In fact if you set `use_hills = 1` it'll take the steep path. Didn't know it's steep, even though "Steig" is quite telling;)

> Maybe valhalla is just not meant for such "workout" like routes

Oh, it absolutely is. At least that's what we strive for. You need to be aware though, that you'll have to set the right costing options (e.g. `use_hills`) to get the desired result. The osm.org version of Valhalla doesn't allow any at all and the osm.de app (ours in fact), does to some degree, but we're also struggling to keep up (PR's welcome;)).

**Comment by nilsnolde** [comment]
One more note to the tagging convention: we can't (and also won't ever, talking for ALL routing engines if I may;)) look at relational context like `natural=wood` or so when deciding on access restrictions for specific countries (you probably see how that gets quickly out of hand). If someone says differently, you can send them to me, then they clearly have little idea how OSM parsing works;) A `vehicle=no` is not very expensive in tagging terms and should be applied for such implicit stuff which likely has little/other applicability in other legislations, in the case where the mapper wants a routing engine to respect it. 

**TLDR**
That being said, I was toying with the idea of having a pre-processor which processes the planet for routers specifically (similar in spirit to the daylight distro, but without any messed up AI). It's a big project though, and even from my POV it'd make a ton of sense, it's not super well received by everyone. I might still go ahead and write a grant application in 2023.

**Comment by hungerburg** [comment]
The api-reference links to a dubious document on the osm-wiki ;) The editor that created the [Austria section](https://wiki.openstreetmap.org/w/index.php?title=OSM_tags_for_routing/Access_restrictions&diff=370828&oldid=363686) is responsible for some of the paths around Poltenhütte! No idea though, why he did not make the track row correct as well, as the path row. Now this will be very hard to find consensus.

Still I think, valhalla [shies from tracks](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.416178941726686%2C47.24946886185169%2C11.415744423866274%2C47.24930135684299%2C11.413008570671082%2C47.248026844862004): With via: 18m up, 9 down, without via: 22 up 13 down, no matter the hills factor? For pedestrians, tracks are not so bad after all.

**Comment by kevinkreiser** [comment]
Yep it seems like we can add a little bit of parsing to get 80% of the stuff we'd want to add wrt pedestrian access. For sure we need to come up with a plan or what is most important to add and in what order

**Comment by hungerburg** [comment]
If I may: 1) Seems that mostly foot=permissive on the access=private footway and driveway does block better directions there. 2) I now explicitly marked Igler Straße as sidewalk=no.

**Comment by hungerburg** [comment]
Not truly on-topic, yet, there is one more curiosity: [The steepest part](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.40584707260132%2C47.24694530635573%2C11.40782117843628%2C47.24685790834034) in the pedestrian route from above looks to be considered flat, while on the ground and also in the altitude shown in the chart it should be ~50%.

**Comment by quaelnix** [comment]
Here, Valhalla takes an illegal exit off the highway:

- https://www.openstreetmap.org/directions?engine=fossgis_valhalla_car&route=51.9352%2C7.5523%3B51.9494%2C7.5549

due to ignoring `vehicle=private` on the following way:

- https://www.openstreetmap.org/way/24544579#map=19/51.94158/7.55196

**Comment by hungerburg** [comment]
In Vienna Convention countries, `vehicle=no` is one of the simplest signs, a red circle with white inner. Sometimes there is an extra add-on sign "including bicycle", though this is just for people with limited knowledge of the signs ;) There can be add-on signs "except forestry" to allow certain groups, that might get mapped `vehicle=agriculture;forestry`. There can also be a picture in the centre, e.g. car/motorcycle/moped/bicycle/&c to prohibit access of certain vehicles only, that would get mapped e.g.. `motor_vehicle=no`. But the vehicle=no sign quasi forms the base for all of them. There is even a version with a pedestrian in centre, that means `foot=no`.

MUTCD regions don't have that, it says at https://wiki.openstreetmap.org/wiki/Key%3Avehicle

**Comment by hungerburg** [comment]
> handle `vehicle=private` the same way that we do for `access=private`

Fine, but foot remains yes there :) Best of course to get multimodal and set bicycle=dismount. Is there a car=park?

**Comment by gknisely** [comment]
Update:  Tested out setting `destination_only_` flag here for this test case.  Seems that the destination only penalty is not high enough to have an impact here.  I believe we need to gather some more test cases and try bumping up the penalty.  We should also retest `access=private` test cases too.  They typically exist around airports.
